src/ophyd_async/core/_device.py

```diff
@@ -1,39 +1,107 @@
+    ):
+        """Connect self and all child Devices.
+
+        Contains a timeout that gets propagated to child.connect methods.
+
+        Parameters
+        ----------
+        mock:
+            If True then use ``MockSignalBackend`` for all Signals
+        timeout:
+            Time to wait before failing with a TimeoutError.
+        """
+
+        # If previous connect with same args has started and not errored, can use it
+        can_use_previous_connect = (
+            mock is self.in_mock_mode
+            and self._task
+            and not (self._task.done() and self._task.exception())
+        )
+        if force_reconnect or not can_use_previous_connect:
+            # Use the connector to make a new connection
+            self.in_mock_mode = mock
+            self._task = asyncio.create_task(
+                self.connect(
+                    mock=mock, timeout=timeout, force_reconnect=force_reconnect
+                )
+            )
+        assert self._task, "Connect task not created, this shouldn't happen"
+        # Wait for it to complete
+        await self._task
+
+    # TODO: we will add some mechanism of invalidating the cache here later
+    @abstractmethod
+    async def connect(
+        self, mock: bool, timeout: float, force_reconnect: bool
+    ) -> None: ...
+
+
+class DeviceChildConnector(DeviceConnector):
+    def __init__(self, device: Device):
+        self._device = device
+
+    async def connect(self, mock: bool, timeout: float, force_reconnect: bool) -> None:
+        coros = {
+            name: child_device.connect(
+                mock=mock, timeout=timeout, force_reconnect=force_reconnect
+            )
+            for name, child_device in self._device.children.items()
+        }
+        await wait_for_connection(**coros)
+
+
+DeviceConnectorType = TypeVar("DeviceConnectorType", bound=DeviceConnector)
+
+
+class Device(HasName, Generic[DeviceConnectorType]):
     """Common base class for all Ophyd Async Devices.
 
     By default, names and connects all Device children.
     """
 
     _name: str = ""
     #: The parent Device if it exists
-    parent: Optional["Device"] = None
-    # None if connect hasn't started, a Task if it has
-    _connect_task: asyncio.Task | None = None
-
-    # Used to check if the previous connect was mocked,
-    # if the next mock value differs then we fail
-    _previous_connect_was_mock = None
+    parent: Device | None = None
+    # The connector to use
+    connect: DeviceConnectorType
 
-    def __init__(self, name: str = "") -> None:
+    def __init__(
+        self,
+        name: str = "",
+        connector: DeviceConnectorType | None = None,
+    ) -> None:
+        if connector is None:
+            # TODO: this is ugly, maybe we remove the option to pass None as the
+            # connector so this goes away?
+            connector = cast(DeviceConnectorType, DeviceChildConnector(self))
+        self.connect = connector
```

Comment:
Apologies, I left a comment before which had an incorrect example - what I meant was

``` python
def foo(device: Device[PviDeviceConnector]):
    assert isinstance(device.connect, DeviceChildConnector)

class DeviceA(Device):
    ...
```
class DeviceB(Device[DeviceChildConnector]):
    ...

device_1: DeviceA = DeviceA()
foo(device_1) # passes type hinting

device_2: DeviceB = DeviceB()
foo(device_2) # correctly fails type hinting
```

in the above code we don't have nice default performance because of this cast.

Reply:
In `class Device`, we need a `connect: DeviceConnector`. In `class Signal(Device)`, we need a `connect: SignalConnector`. To make this compatible we make `class Device(Generic[DeviceConnectorT])`, and make `class Signal(Device[SignalConnector])`. However, `Device.__init__` takes `DeviceConnectorT | None`. If we supply a `DeviceConnectorT` then typing is happy, but if we supply `None` then typing cannot infer its type, so when we create it in the code above, it is of the wrong type (`DeviceChildConnector`, not `DeviceConnectorT`). This is a valid concern, typing does not know that `Signal` will always pass `SignalConnector` in the constructor rather than `None`. I'm not sure what to do about this concern, and I'm not sure if there's anything in your comment that will fix this particular concern? Also, at runtime `typing.cast` [does nothing](https://github.com/python/cpython/blob/cceb25c5033bce5d19311d63f6943215a1a27283/Lib/typing.py#L2182-L2190) so there is no performance hit.

Reply:
Here's one solution if we want to forbid subclasses of types without a Connector in the generic e.g `class A(Device):`

``` python
from typing import Generic, TypeVar, get_args, cast
import pytest

class SomeClass:
    def __init__(self, base: "Base"): ...

class SomeClass1(SomeClass): ...

class SomeClass2(SomeClass): ...

T = TypeVar("T", bound=SomeClass)

class Base(Generic[T]):
    value: T

    def __init__(self, value: T | None = None):
        # In 3.12 you can use types.get_original_bases
        annotation = type(self).__orig_bases__[0] # pyright: ignore
        print(type(self).__name__, "ANNOTATED", annotation)
        arg = get_args(annotation)
        print(type(self).__name__, "ARGS", annotation)
        type_in_generic = arg[0]
        if type_in_generic is T:
            # This means they didn't provide a type
            raise RuntimeError

        print(type(self).__name__, "TYPE IN GENERIC", type_in_generic)

        # Type hinting would complain if this wasn't true as well
        assert issubclass(type_in_generic, SomeClass)

        value = value or cast(T, type_in_generic(self))

class Sub1(Base[SomeClass1]): ...

class Sub2(Base[SomeClass2]): ...

class Sub3(Base): ...

def foo(x: Base[SomeClass1]) -> None: ...


sub1 = Sub1()
foo(sub1) # This correctly passes type checking

sub2 = Sub2()
foo(sub2) # This correctly fails type checking

with pytest.raises(RuntimeError):
    sub3 = Sub3() # This can't even be initialised
```

With output:
```
Sub1 ANNOTATED __main__.Base[__main__.SomeClass1]
Sub1 ARGS __main__.Base[__main__.SomeClass1]
Sub1 TYPE IN GENERIC <class '__main__.SomeClass1'>
Sub2 ANNOTATED __main__.Base[__main__.SomeClass2]
Sub2 ARGS __main__.Base[__main__.SomeClass2]
Sub2 TYPE IN GENERIC <class '__main__.SomeClass2'>
Sub3 ANNOTATED typing.Generic[~T]
Sub3 ARGS typing.Generic[~T]
```

This is very hacky (until python 3.12), but it means that Devices require the connector.